src/utils/SubscribableRequestQueue.ts

```diff
@@ -0,0 +1,175 @@
+import RequestQueue from "./RequestQueue";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Resolver = (value?: any) => void;
+type Rejecter = (reason?: unknown) => void;
+
+type RequestSubscription = {
+  subscriberId: number;
+  resolve: Resolver;
+  reject: Rejecter;
+};
+
+/**
+ * An extension of `RequestQueue` that adds a concept of "subscribers," which may share references to a single request
+ * or cancel their subscription without disrupting the request for other subscribers.
```

Comment:
just to refresh my memory, does this mean that each subscriber gets a unique promise instead of having one shared promise for each unique data request?

Reply:
Right, this class owns a `RequestQueue` and that underlying queue still creates promises that run a task and resolve when complete. This class is just about managing a set of subscriptions (which are themselves promises) to those promises and saying "when the queued promise resolves with a value, resolve all the subscribing promises with that value."